healthcheck.go

```diff
@@ -0,0 +1,117 @@
+package hc
+
+import (
+	aah "aahframe.work"
+	"sync"
+	"time"
+)
+
+// HealthCheckReporter contains the dependencies to check
+type HealthCheckReporter struct {
+	dependencies    map[string]Dependency
+	globalHealth    bool
+	globalHealthMsg map[string]string
+	mu              sync.RWMutex
+}
+
+// HealthChecker is the interface for a Dependency that can be health-checked
+type HealthChecker interface {
+	// Check will return nil if dependency is reachable/healthy
+	// You should implement this func with a sensible timeout (< 3 or 5 sec)
+	Check() error
+}
+
+// Dependency implements the HealthChecker interfance and also contains the check config
+type Dependency struct {
+	// implements HealthChecker interface
+	HealthChecker
+	// softDep - if true it will allow errors so won't report unhealthy
+	softDep bool
+	// TODO: interval unused for now - got a global interval for all checks
+	interval time.Duration
+}
+
+// NewHealthCheckReporter returns an empty HealthCheckReporter
+func NewHealthCheckReporter() *HealthCheckReporter {
+	reporter := &HealthCheckReporter{
+		dependencies: make(map[string]Dependency),
+		globalHealth: true,
+	}
+	go func() {
+		t := time.NewTicker(5 * time.Second)
+		for {
+			<-t.C
+			reporter.RunChecks()
+		}
+	}()
+
+	return reporter
+}
+
+// RegisterDependency adds a dependency to check it periodically
+func (hr *HealthCheckReporter) RegisterDependency(name string, dep Dependency) {
+	hr.mu.Lock()
+	defer hr.mu.Unlock()
+	_, ok := hr.dependencies[name]
+	if ok {
+		panic("Dependency already exists: " + name)
+	}
+	hr.dependencies[name] = dep
+}
+
+// RunChecks performs a check in all the dependencies and update the global status
+func (hr *HealthCheckReporter) RunChecks() {
```

Comment:
could unexport this function so only this pkg has the ability to run the checks

Reply:
Yes, unexported would be appropriate here 👍  Typical design thoughts, it's good to start with less exported options so that we prevent breaking changes.